### PR TITLE
python@3.9: build --with-system-libmpdec

### PR DESCRIPTION
**Pull request by skrah**
- [ x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by skrah** [comment]
This builds Python with the `homebrew` `libmpdec`.  I've tested the build in verbose mode on `linuxbrew` and all looks well.

The trivial patch is required due to small header cleanups in `mpdecimal.h` in `mpdecimal-2.5.1`.  Debian bullseye, Arch Linux,  Alpine Linux etc. also carry this patch.

**Comment by Bo98** [comment]
This seems reasonable. Appears to not be working on macOS though, judging by the CI.

Regarding the patch: is there plans to upstream this at some point?

**Comment by sharris40** [comment]
Here are the relevant errors from the uploaded logs:

`/private/tmp/python-3.9-20210223-43981-1jeibdp/Python-3.9.2/Modules/_decimal/_decimal.c:4523:6: error: "No valid combination of CONFIG_64, CONFIG_32 and _PyHASH_BITS"`
`#error "No valid combination of CONFIG_64, CONFIG_32 and _PyHASH_BITS"`

`/private/tmp/python-3.9-20210223-43981-1jeibdp/Python-3.9.2/Modules/_decimal/_decimal.c:4570:43: error: use of undeclared identifier 'p'`
`mpd_qpowmod(exp_hash, &ten, tmp, &p, &maxctx, &status);`

`/private/tmp/python-3.9-20210223-43981-1jeibdp/Python-3.9.2/Modules/_decimal/_decimal.c:4575:32: error: use of undeclared identifier 'inv10_p'`
`mpd_qpowmod(exp_hash, &inv10_p, tmp, &p, &maxctx, &status);`

`/private/tmp/python-3.9-20210223-43981-1jeibdp/Python-3.9.2/Modules/_decimal/_decimal.c:4575:47: error: use of undeclared identifier 'p'`
`mpd_qpowmod(exp_hash, &inv10_p, tmp, &p, &maxctx, &status);`

`/private/tmp/python-3.9-20210223-43981-1jeibdp/Python-3.9.2/Modules/_decimal/_decimal.c:4590:25: error: use of undeclared identifier 'p'`
`mpd_qrem(tmp, tmp, &p, &maxctx, &status);`

**Comment by skrah** [comment]
Yes, the patch will probably end up in upstream at some point.

Thanks for the logs, I couldn't find this part.  The issue is that Python ships `universal` installers on `macOS`, so `setup.py` contains this part:

```
        elif MACOS:
            # Universal here means: build with the same options Python
            # was built with.
            define_macros = config['universal']
```

However, using it means that `mpdecimal` would need to be built with a universal `mpdecimal.h`, i.e. using ` ./configure MACHINE=universal`.  So there are at least two choices:

1) If people never build or use universal libraries on `homebrew`, then perhaps it's best to delete that passage from `setup.py` and add it to the patch (this change cannot be upstreamed).

2) Otherwise, build `mpdecimal` with the universal header, which may turn out messy for `linuxbrew`, though I think it should work if the only supported platform is `amd64`.

**Comment by skrah** [comment]
Actually, if exporting environment variables is acceptable for `homebrew`:

3. Export the environment variable `PYTHON_DECIMAL_WITH_MACHINE=x64` (`uint128` for `aarch64`, `ppro` for `x86`).  No further patch would be needed.

**Comment by Bo98** [comment]
> 3. Export the environment variable `PYTHON_DECIMAL_WITH_MACHINE=x64` (`uint128` for `aarch64`, `ppro` for `x86`).  No further patch would be needed.

That would be fine. There's many cases where we do `ENV["key"] = "value"` in builds.

We don't support 32-bit so it's just `x86_64` and `arm64/aarch64` to handle (`if Hardware::CPU.arm?` or similar). We also do not build universal binaries.

File: Formula/python@3.9.rb [hunk 1]
```diff
@@ -102,6 +110,11 @@ def install
     ENV["PYTHONHOME"] = nil
     ENV["PYTHONPATH"] = nil
 
+    # Override the auto-detection in setup.py, which assumes a universal build.
+    if OS.mac?
```
Comment (Bo98) [inline Formula/python@3.9.rb:114]: Minor nitpick:

```suggestion
    on_macos do
```

This is more consistent with what other formulae do (albeit functionally the same).

**Comment by Bo98** [comment]
Looks good otherwise - it seems to be working :+1:

File: Formula/python@3.9.rb (hunk 1 above)
Comment (skrah) [reply Formula/python@3.9.rb]: Indeed, that looks nicer.

**Comment by skrah** [comment]
The tests pass on `arm`, and the other failures look unrelated (`llvm` etc.).

**Comment by Bo98** [comment]
Yeah, that can happen when testing so much in one go. (That disk space issue is a little concerning though - I'll need to investigate that machine.)

Thanks for the contribution @skrah, I'll merge this now.

**Comment by Bo98** [comment]
Merged, with thanks: 6af4420d9c033d08fe000049cb5673016cd49edf, 578501f1556b739170edf4c991ca12a49ac82616

**Comment by skrah** [comment]
Thank you!

**Pull request closed without merging**